proposals/storage-get-keys.md

```diff
@@ -0,0 +1,115 @@
+# Proposal: browser.storage.&lt;area>.getKeys
+
+**Summary**
+
+API to retrieve all keys for a given storage area in the `browser.storage` API.
+
+**Document Metadata**
+
+**Author:** [Oliver Dunk (Google)](https://github.com/oliverdunk)
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** [polywock](https://github.com/polywock)
+
+**Created:** 2024-06-28
+
+**Related Issues:** [#601](https://github.com/w3c/webextensions/issues/601)
+
+## Motivation
+
+### Objective
+
+This API enables developers to retrieve the keys in storage without needing to
+receive all of the data stored alongside those keys. Depending on the
+implementation, the impact of this optimization may differ - for example
+Chromium has only a limited ability to access data from storage without
+fetching the associated data. Regardless, this reduces the amount of data
+returned to the caller and allows for further optimizations in the future.
+
+#### Use Cases
+
+There are existing use cases for enumerating the data in extension storage. For
+example, extensions for Chrome that add extension storage areas to DevTools.
+While these also need a way to show the associated values, they may not want to
+load these upfront.
+
+Additionally, in some cases developers may want to access keys with a given
+prefix, for example if using a key like `foo:tabid`. This API proposal does not
+allow matching by prefix (see "Future Work") but developers would be able to do
+this themselves after obtaining the list of keys.
+
+### Known Consumers
+
+There is developer interest for this API in
+[#601](https://github.com/w3c/webextensions/issues/601). While we don't expect
+significant immediate usage, it is a small, well-scoped API and should not
+represent significant engineering effort.
+
+## Specification
+
+### Schema
+
+```ts
+interface StorageArea {
+    getKeys(): Promise<string[]>;
+}
+```
+
+The original issue proposed the `getAllKeys()` name, to match IndexedDB. This
+proposal suggests the slightly shorter `getKeys()`, since that seems equally
+clear and less verbose.
+
+### Behavior
+
+This API will throw / populate browser.runtime.lastError in the case of
```

Comment:
`This API will reject in the case of…` since there is no callback version, right?

(However, in Safari / WebKit, everything that returns a promise can take a final callback parameter instead, even in v3 for v2 compatibility. Maybe we should change that…)

Reply:
That's a good question - I intentionally didn't focus too much on callbacks for the proposal, since as you mentioned they are mostly for backwards compatibility. Perhaps a brief mention is still worthwhile though - we should definitely support them.

Reply:
```suggestion
This API will reject with an Error / populate browser.runtime.lastError in the case of
```
I don't expect any error to be thrown, unless invalid parameters are passed. The normal mode of error handling here is by rejecting the Promise.

Reply:
Good spot, thanks. That's what I was trying to say.